airflow/cli/api/client.py

```diff
@@ -0,0 +1,226 @@
+    PoolsOperations,
+    ProvidersOperations,
+    ServerResponseError,
+    VariablesOperations,
+    VersionOperations,
+)
+from airflow.version import version
+
+if TYPE_CHECKING:
+    # # methodtools doesn't have typestubs, so give a stub
+    def lru_cache(maxsize: int | None = 128):
+        def wrapper(f):
+            return f
+
+        return wrapper
+else:
+    from methodtools import lru_cache
+
+log = structlog.get_logger(logger_name=__name__)
+
+__all__ = [
+    "Client",
+    "Credentials",
+]
+
+
+def add_correlation_id(request: httpx.Request):
+    request.headers["correlation-id"] = str(uuid7())
+
+
+def get_json_error(response: httpx.Response):
+    """Raise a ServerResponseError if we can extract error info from the error."""
+    err = ServerResponseError.from_response(response)
+    if err:
+        log.warning("Server error ", extra=dict(err.response.json()))
+        raise err
+
+
+def raise_on_4xx_5xx(response: httpx.Response):
+    return get_json_error(response) or response.raise_for_status()
+
+
+# Credentials for the API
+class Credentials:
+    """Credentials for the API."""
+
+    api_url: str | None
+    api_token: str | None
+    api_environment: str
+
+    def __init__(
+        self,
+        api_url: str | None = None,
+        api_token: str | None = None,
+        api_environment: str = "production",
+    ):
+        self.api_url = api_url
+        self.api_token = api_token
+        self.api_environment = os.getenv("APACHE_AIRFLOW_CLI_ENVIRONMENT") or api_environment
+
+    @property
+    def input_cli_config_file(self) -> str:
+        """Generate path and always generate that path but let's not world readable."""
+        return f"{self.api_environment}.json"
+
+    def save(self):
+        """Save the credentials to keyring and URL to disk as a file."""
+        default_config_dir = user_config_path("airflow", "Apache Software Foundation")
+        if not os.path.exists(default_config_dir):
+            os.makedirs(default_config_dir)
+        with open(os.path.join(default_config_dir, self.input_cli_config_file), "w") as f:
+            json.dump({"api_url": self.api_url}, f)
+        keyring.set_password("airflow-cli", f"api_token-{self.api_environment}", self.api_token)
+
+    def load(self) -> Credentials:
+        """Load the credentials from keyring and URL from disk file."""
+        default_config_dir = user_config_path("airflow", "Apache Software Foundation")
+        if os.path.exists(default_config_dir):
+            with open(os.path.join(default_config_dir, self.input_cli_config_file)) as f:
+                credentials = json.load(f)
+                self.api_url = credentials["api_url"]
+                self.api_token = keyring.get_password("airflow-cli", f"api_token-{self.api_environment}")
+            return self
+        else:
+            rich.print("[red]No credentials found.")
+            rich.print("[green]Please run: [blue]airflow auth login")
+            sys.exit(1)
```

Comment:
We shouldn't be printing here as this makes it harder to use this "properly". This should raise something, and yhe print and exit be closer to the cli layer

Reply:
Moved them to the CLI layer, and everything is now just throwing the exceptions